### PR TITLE
Fix: setContext doesn't trigger componentWillReceiveProps

### LINKED ISSUE #2258: setContext doesn't trigger componentWillReceiveProps
### Current behavior

`setContext` doesn't trigger `componentWillReceiveProps`

### Expected behavior

`setContext` should trigger `componentWillReceiveProps`.

#### API

- [x] shallow
- [ ] mount
- [ ] render

#### Version

| library             | version
| ------------------- | -------
| enzyme              |  3.10.0
| react               | 16.3.18
| react-dom           | 16.3.3
| react-test-renderer | 16.10.2
| adapter (below)     | 1.14.0

#### Adapter

- [x] enzyme-adapter-react-16
- [ ] enzyme-adapter-react-16.3
- [ ] enzyme-adapter-react-16.2
- [ ] enzyme-adapter-react-16.1
- [ ] enzyme-adapter-react-15
- [ ] enzyme-adapter-react-15.4
- [ ] enzyme-adapter-react-14
- [ ] enzyme-adapter-react-13
- [ ] enzyme-adapter-react-helper
- [ ] others ( )

### PR DESCRIPTION
**Pull request by mul53**
Fixes #2258

Below are two code sandboxes that show the bug exists, i added enzyme to the sandbox when you run the tests sandbox 2 doesn't call the method while sandbox 1 does.
-  [Sandbox #1](https://codesandbox.io/s/distracted-sun-8vqh9) enzyme 3.3.0, react 16.14.0
-  [Sandbox #2](https://codesandbox.io/s/lucid-wilbur-wkotf) enzyme 3.4.0, react 16.0.0

### CODE REVIEW

**Comment by mul53** [comment]
@ljharb should i also add the patch for adapter versions 16.1-3

**Comment by ljharb** [comment]
@mul53 yep! however, if this PR is otherwise good to go, I'll go ahead and backport the fixes for you; if not, I'll leave it for you :-)

(also, in the future, you may not want to make PRs from master, so things are easier on other fellows making PRs from the same fork, but nbd)

**Review by ljharb — APPROVED** [review]
Thanks, this looks great!

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 1]
```diff
@@ -2122,34 +2122,20 @@ describe('shallow', () => {
...
-            expect(spy.args).to.deep.equal([
-              ['componentWillMount'],
-              ['render'],
-            ]);
-            spy.resetHistory();
             wrapper.setContext({ foo: 'foo' });
+
             expect(spy.args).to.deep.equal([
+              // TODO: figure out what this test should actually be
+              ...(is('< 16 || ~16.3') ? [['componentWillReceiveProps']] : []),
```
Comment (ljharb) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx:2138]: even with this, it's failing in 16.3 with the 16 adapter - but if i remove the `~16.3` from here, it fails with the 16.3 adapter.

@mul53 could you try to make more sandboxes with react 13, 14, 15, 16.2, 16.3, and 16.4, to see what this test should actually be doing?

Comment (mul53) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx:2138]: okay i will do that, should have them ready by tomorrow

**Comment by mul53** [comment]
Here are the sandboxes

[React 0.14 Sandbox](https://codesandbox.io/s/setcontext-react-014-jhwel)
[React 15 Sandbox](https://codesandbox.io/s/setcontext-react-15-wirdb)
[React 16.2 Sandbox](https://codesandbox.io/s/setcontext-react-162-wg51l)
[React 16.3 Sandbox](https://codesandbox.io/s/setcontext-react-163-n757b)
[React 16.4 Sandbox](https://codesandbox.io/s/setcontext-react-164-ww626)

**Comment by ljharb** [comment]
hmm, that was not intentional

**Comment by ljharb** [comment]
@mul53 i think you'll need to create a new PR from 63a40541b5c34e689633428fde2599c925ba56e5 that's from a branch that's not `master`, since the MLH fork has the pull bot set up on it that's messing things up :-/

**Comment by mul53** [comment]
@ljharb I will do so, there are tests to fix

**Pull request merged**